### PR TITLE
C++: Add barriers to `cpp/uncontrolled-allocation-size`

### PR DESCRIPTION
**Pull request by MathiasVP**
(Part of https://github.com/github/codeql-c-team/issues/272.)

This PR adds two barriers to `cpp/uncontrolled-allocation-size`:

- The first barrier (added in 5031b73f3509e4d1c1ce21586dd1262502177670) blocks flow out of arithmetic operations that we can prove won't overflow. This removes false positives in a couple of LGTM projects. When I compare [5031b73](https://lgtm.com/query/4264381093598795328/) to [main](https://lgtm.com/query/6134533116057054671/) we get:

| Project | Alerts with main | Alerts with 5031b73 |
|--------|--------|--------|
| openssl/openssl | 8 | 7 |
| gpac/gpac | 2 | 0 |
| libretro/RetroArch | 2 | 1 |
| axiomatic-systems/Bento4 | 1 | 0 |
| systemd/systemd | 11 | 9 | 
| rizinorg/cutter | 33 | 25 | 
| radareorg/radare2 | 29 | 22 | 
| haiku/haiku | 14 | 12 | 
| kernel.org/pub/scm/network/iproute2/iproute2-next | 3 | 1 | 
| Chia-Network/chiapos | 1 | 0 | 
| ArtifexSoftware/ghostpdl | 3 | 1 | 

- The second barrier (added in 2d0a56128d8f10acade04f99c73493671186d298) blocks flow out of `PointerDiffExpr`. This is quite a common cause of false positives for this query. When I compare [5031b73](https://lgtm.com/query/4264381093598795328/) to [2d0a561](https://lgtm.com/query/6799893794923674095/)) we get:

| Project | Alerts with 5031b73 | Alerts with 2d0a561 |
|--------|--------|--------|
vim/vim | 1 | 0
openssl/openssl | 7 | 0
git/git | 1 | 0
lxc/lxc | 1 | 0
alibaba/AliSQL | 1 | 0
apache/trafficserver | 2 | 1
stellar/stellar-core | 3 | 0
systemd/systemd | 9 | 6
nmap/nmap | 1 | 0
cfengine/core | 2 | 1
rizinorg/cutter | 25 | 24
haiku/haiku | 12 | 10
horms/openvswitch | 1 | 0
trini/u-boot | 1 | 0
FreeRADIUS/freeradius-server | 1 | 0
ArtifexSoftware/ghostpdl | 1 | 0

This change doesn't seem to affect SAMATE.

### CODE REVIEW

File: cpp/ql/test/query-tests/Security/CWE/CWE-190/semmle/TaintedAllocationSize/test.cpp [hunk 1]
```diff
@@ -139,7 +139,7 @@ void more_bounded_tests() {
 
 		if (size > 0)
 		{
-			malloc(size * sizeof(int)); // BAD
+			malloc(size * sizeof(int)); // GOOD (overflow not possible)
```
Comment (rdmarsh2) [inline cpp/ql/test/query-tests/Security/CWE/CWE-190/semmle/TaintedAllocationSize/test.cpp:142]: Could we add a test with a `long` to keep testing `size > 0` with an actual overflow?

Comment (MathiasVP) [reply cpp/ql/test/query-tests/Security/CWE/CWE-190/semmle/TaintedAllocationSize/test.cpp:143]: Good point. I've done this in 31091c66c19fdb401967decc596621fa68db8d0a. Sorry about the horrible diff from that commit. (I've checked that it does the right thing.)

File: cpp/ql/src/Security/CWE/CWE-190/TaintedAllocationSize.ql [hunk 2]
```diff
@@ -42,6 +42,12 @@ class TaintedAllocationSizeConfiguration extends TaintTrackingConfiguration {
       e instanceof AssignArithmeticOperation
     ) and
     not convertedExprMightOverflow(e)
+    or
+    // Subtracting two pointers is either well-defined (and the result will likely be small), or
+    // terribly undefined and dangerous. Here, we assume that the programmer has ensured that the
+    // result is well-defined (i.e., the two pointers point to the same object), and thus the result
+    // will likely be small.
```
Comment (geoffw0) [inline cpp/ql/src/Security/CWE/CWE-190/TaintedAllocationSize.ql:49]: Seems reasonable.  If the subtraction result is not well-defined, that should probably be handled by another query (which may or may-not already exist) rather than blaming a later `malloc`.

**Review by geoffw0 — APPROVED** [review]
LGTM.

File: cpp/ql/src/Security/CWE/CWE-190/TaintedAllocationSize.ql (hunk 2 above)
Comment (MathiasVP) [reply cpp/ql/src/Security/CWE/CWE-190/TaintedAllocationSize.ql:49]: Indeed. 👍 I've opened up https://github.com/github/codeql-c-team/issues/525 for this exact query idea.
